### PR TITLE
Fugue Integration

### PR DESCRIPTION
**Pull request by goodwanghan**
## Description

This repository needs to integrate with [Fugue](https://github.com/fugue-project/fugue) so whyllogs profiling can work on Spark, Dask and Ray agnostically.

The commits in this pull request will:
1. Finish the initial implementation of profiling with/without logical partitioning
2. Docs and instructions

- [x] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -15,7 +15,7 @@ python = ">=3.7.1, <4"
 whylogs-sketching = ">=3.4.1.dev3"
 protobuf = ">=3.15.5"
 importlib-metadata = { version = ">=0.23", python = "<3.8" }
-typing-extensions = ">=typing-extensions"
+# typing-extensions = ">=typing-extensions"
```
Comment (goodwanghan) [inline python/pyproject.toml]: @andyndang 👀

Comment (andyndang) [reply python/pyproject.toml]: @jamie256 👀?

Comment (TheMellyBee) [reply python/pyproject.toml]: Looks like the original was put in by @andyndang 5 months ago. I had readthedocs poetry fail on friday on this, but it appeared to be fixed by what @jamie256. I'm super curious what this is for?

File: python/examples/integrations/Fugue_Profiling.ipynb [hunk 2]
```diff
@@ -0,0 +1,1162 @@
...
+       "DataFrame[a: bigint, b: string, __whylogs_df_profile_view: binary]"
+      ]
+     },
+     "execution_count": 23,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "profile(tdf, partition={\"by\":[\"a\",\"b\"]}, engine=spark, as_local=False) # returns a native pyspark dataframe"
```
Comment (TheMellyBee) [inline python/examples/integrations/Fugue_Profiling.ipynb]: What is the bad pipe message?

File: python/examples/integrations/Pyspark_Profiling.ipynb [hunk 3]
```diff
@@ -659,7 +659,7 @@
    "hash": "16a10773934acde374a1cd808bcd53b1085f60e17ec18f4c0c26564dd890a5a0"
   },
   "kernelspec": {
-   "display_name": "Python 3.9.12 ('.venv': venv)",
+   "display_name": "Python 3 (ipykernel)",
```
Comment (TheMellyBee) [inline python/examples/integrations/Pyspark_Profiling.ipynb:662]: @jamie256 is this something we want to change for everyone?

File: python/examples/integrations/Dask_Profiling.ipynb [hunk 4]
```diff
@@ -1555,7 +1555,7 @@
  ],
  "metadata": {
   "kernelspec": {
-   "display_name": "Python 3.9.13 ('.venv': poetry)",
+   "display_name": "Python 3 (ipykernel)",
```
Comment (TheMellyBee) [inline python/examples/integrations/Dask_Profiling.ipynb:1558]: @jamie256 same as another comment, is this something we want changed for everyone?

File: python/docs/datasets [hunk 5]
```diff
@@ -0,0 +1 @@
+../whylogs/datasets/descr
```
Comment (TheMellyBee) [inline python/docs/datasets:1]: @goodwanghan Could you clarify the purpose of this? :)

File: python/examples/integrations/Fugue_Profiling.ipynb [hunk 6]
```diff
@@ -0,0 +1,1162 @@
+{
+ "cells": [
```
Comment (TheMellyBee) [inline python/examples/integrations/Fugue_Profiling.ipynb:2]: Can you quiet the block output to make this cleaner and easier to read? (won't show object from the last line each block)

File: python/whylogs/api/fugue/profiler.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    def transform_with_logical_partition(self, df: Any, **kwargs: Any) -> Any:
+        return transform(
+            df,
+            transformer(lambda pdf: pdf.schema.extract(self._by) + self._profile_schema)(self.profile_partition),
+            partition=self._partition,
+            **kwargs,
+        )
+
+
+def profile(
```
Comment (TheMellyBee) [inline python/whylogs/api/fugue/profiler.py]: I'd prefer a less generic name to make this clear that this is for Fugue. Especially as this is in the API. How about something like FugueDatasetProfile or FugueProfiler?

File: python/whylogs/api/fugue/profiler.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+from whylogs.core.view.column_profile_view import ColumnProfileView
+from whylogs.core.view.dataset_profile_view import DatasetProfileView
+
+COL_NAME_FIELD = "__whylogs_fugue_col_name"
+COL_PROFILE_FIELD = "__whylogs_fugue_col_profile"
+COL_PROFILE_SCHEMA = Schema([(COL_NAME_FIELD, str), (COL_PROFILE_FIELD, bytes)])
+DF_PROFILE_FIELD = "__whylogs_df_profile_view"
+
+
+class _Profiler:
```
Comment (TheMellyBee) [inline python/whylogs/api/fugue/profiler.py]: Same as comment below on the name, but FugueDatasetPofiler would be more descriptive

**Review by TheMellyBee — COMMENTED** [review]
I have a couple of questions and some comments around cleaning it up and clarity. Over all though it's looking in a good direction.

File: python/pyproject.toml (hunk 1 above)
Comment (jamie256) [reply python/pyproject.toml]: This is for backporting some of the language features we used in whylogs v1 that are only in python 3.10, so we only need this for older than python 3.10. Our use of TypeAlias in the decorators is one example.

(Attempting to) fix the version range here: https://github.com/whylabs/whylogs/pull/845

Comment (jamie256) [reply python/pyproject.toml]: Ok, the pyproject.toml file is now updated in mainline with https://github.com/whylabs/whylogs/pull/845 merged.

File: python/examples/integrations/Pyspark_Profiling.ipynb (hunk 3 above)
Comment (jamie256) [reply python/examples/integrations/Pyspark_Profiling.ipynb:662]: It's fine, the kernelspec display_name will pick up the last kernel the notebook was run on if you are editing and testing it. The previous value wasn't intentional so much as incidental to the developer workflow of the previous authors.

But I'm guessing this file doesn't need any edits for this PR?

File: python/whylogs/api/fugue/profiler.py (hunk 7 above)
Comment (jamie256) [reply python/whylogs/api/fugue/profiler.py]: The method would still need to be imported from whylogs.api.fugue which helps clarify this is in context of a fugue integration, so I think it is ok.

File: python/whylogs/api/fugue/profiler.py (hunk 8 above)
Comment (jamie256) [reply python/whylogs/api/fugue/profiler.py]: underscore prefixed names within a integration's folder are ok to be somewhat generic, its not part of whylogs.core where I would want more differentiation.

**Review by jamie256 — COMMENTED** [review]
This approach looks really nice to integrate whylogs and fugue! Let me know how I can help. 

This looks close, I think with some minor cleanup if you have your example notebook working, and the CI checks passing we can merge this? I will check tomorrow morning, and happy to help resolve issues.

**Comment by goodwanghan** [comment]
Thank @jamie256 the notebook is working. I will also add some comments on the public function. Then we are good to merge. Currently I am having a hard time with poetry I am hitting this issue https://github.com/python-poetry/poetry/issues/4242 and may try poetry 1.2 which fixes it.

**Comment by jamie256** [comment]
> I am hitting this issue [python-poetry/poetry#4242](https://github.com/python-poetry/poetry/issues/4242) and may try poetry 1.2 which fixes it.

I updated poetry to version 1.2.1 in our CI configs as it now works after fixing the typing-extensions entry: https://github.com/whylabs/whylogs/pull/847

**Comment by goodwanghan** [comment]
> > I am hitting this issue [python-poetry/poetry#4242](https://github.com/python-poetry/poetry/issues/4242) and may try poetry 1.2 which fixes it.
> 
> I updated poetry to version 1.2.1 in our CI configs as it now works after fixing the typing-extensions entry: #847

Thanks, but please let me do that, I am making a couple of changes

**Comment by goodwanghan** [comment]
Merging poetry.lock is not fun 😢

**Comment by jamie256** [comment]
> Merging poetry.lock is not fun 😢

Looks like you have it working! And notebooks cleaned up.

`poetry.lock` should be generated from pyproject.toml file changes by running `poetry update` (if possible), manually updating the lock file is painful.

File: python/tests/api/pyspark/experimental/test_profiler_function.py [hunk 9]
```diff
@@ -55,7 +55,7 @@ def test_profile_mapper_function(self, input_df):
             column_profiles.collect(),
         )
         assert profile_bytes_df.rdd.getNumPartitions() > 0
-        assert profile_bytes_df.count() == 4
+        assert profile_bytes_df.count() > 0
```
Comment (jamie256) [inline python/tests/api/pyspark/experimental/test_profiler_function.py:58]: Yeah thanks, this was a sloppy assert.

File: python/whylogs/api/fugue/__init__.py [hunk 10]
```diff
@@ -0,0 +1,2 @@
+# flake8: noqa
+from .profiler import fugue_profile
```
Comment (jamie256) [inline python/whylogs/api/fugue/__init__.py:2]: thanks for renaming, looks better.

**Review by jamie256 — APPROVED** [review]
Thanks for the push and all the updates here, I'll pull this into today's release.

**Review by andyndang — APPROVED** [review]
lgtm!

**Pull request merged**